File: src/saga-desktop/src/net/wombatrpgs/saga/rpg/mutant/MutationManager.java
/**
 *  MutantLevelManager.java
 *  Created on May 24, 2014 8:31:41 AM for project saga-desktop
 *  Author: psy_wombats
 *  Contact: psy_wombats@wombatrpgs.net
 */
package net.wombatrpgs.saga.rpg.mutant;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import net.wombatrpgs.mgne.core.MGlobal;
import net.wombatrpgs.saga.core.SGlobal;
import net.wombatrpgs.saga.rpg.battle.Battle;
import net.wombatrpgs.saga.rpg.chara.Chara;
import net.wombatrpgs.sagaschema.rpg.stats.Stat;

/**
 * Object that a mutant gets to manage what stats or levels it gets as a result
 * of battle by recording what happens to the mutant.
 */
public class MutationManager {
	
	protected static final int MHP_GAIN_MIN = 5;
	protected static final int MHP_GAIN_MAX = 19;
	
	protected Chara chara;
	protected Battle battle;
	protected int level;
	
	protected Map<MutantEvent, Integer> hits;
	
	/**
	 * Creates a new manager for a character. Note that the character should be
	 * a mutant and that the different races should probably be subclasses
	 * rather than an enum, oh well. Only specific to a single battle.
	 * @param	battle			The battle this manager records for
	 * @param	chara			The character to create for
	 */
	public MutationManager(Chara chara, Battle battle) {
		this.chara = chara;
		this.battle = battle;
		hits = new HashMap<MutantEvent, Integer>();
		level = battle.getLevel();
	}
	
	/**
	 * Records an event as having occurred.
	 * @param	event			The event that occurred
	 */
	public void recordEvent(MutantEvent event) {
		Integer existing = hits.get(event);
		if (existing == null) existing = 0;
		hits.put(event, existing + 1);
	}
	
	/**
	 * Finalize the event list and produce a list of available level up options.
	 * This should always be length 2 unless something changes? Should only be
	 * called once per instance.
	 * @return					A list of possible leveling options
	 */
	public List<Mutation> produceOptions() {
		List<MutantEvent> occurred = new ArrayList<MutantEvent>();
		for (MutantEvent event : hits.keySet()) {
			if (hits.get(event) > 0) occurred.add(event);
		}
		Mutation fixed = null;
		if (occurred.size() > 0) {
			switch (occurred.get(MGlobal.rand.nextInt(occurred.size()))) {
			case DAMAGED:
				fixed = SGlobal.settings.getMutations().genHealthOption(chara);
				break;
			case DAMAGED_PHYSICALLY:
				fixed = new MutationStat(chara, Stat.DEF);
				break;
			case USED_ABILITY:
				fixed = new MutationAbil(chara, level);
				break;
			case USED_AGI:
				fixed = new MutationStat(chara, Stat.AGI);
				break;
			case USED_MANA:
				fixed = new MutationStat(chara, Stat.MANA);
				break;
			case USED_STR:
				fixed = new MutationStat(chara, Stat.STR);
				break;
			}
		} else {
			fixed = SGlobal.settings.getMutations().randomOption(chara, level);
		}
		
		List<Mutation> results = new ArrayList<Mutation>();
		while (isMaxed(fixed)) {
			fixed = SGlobal.settings.getMutations().randomOption(chara, level);
		}
		results.add(fixed);
		
		for (int tries = 0; tries < 100; tries += 1) {
			Mutation random = SGlobal.settings.getMutations().randomOption(chara, level);
			if (fixed.getStat() != random.getStat() && !isMaxed(random)) {
				results.add(random);
				break;
			}
		}
		if (results.size() < 2) {
			results.add(new MutationAbil(chara, level));
		}
		
		return results;
	}
	
	/**
	 * Checks if a mutation has reached max stat.
	 * @param	mutation		The mutation to check
	 * @return					True if it is
	 */
	protected boolean isMaxed(Mutation mutation) {
		Stat stat = mutation.getStat();
		if (stat != null) {
			if (stat == Stat.MHP) {
				return chara.get(stat) >= 999;
			} else {
				return chara.get(stat) >= 99;
			}
		} else {
			return false;
		}
	}

}

File: saga-unity/Assets/Scripts/RPG/Mutant/MutationManager.cs
using UnityEngine;
using System.Collections.Generic;

public class MutationManager : MonoBehaviour {

    protected const int MhpGainMin = 5;
    protected const int MhpGainMax = 19;

    protected Unit unit;
    protected Battle battle;
    protected int level;

    protected Dictionary<MutantEvent, int> hits;

    private const string MutationTablePath = "Database/MutationSettings/mutations_default";
    private static MutationSettingsData mutationsTable;
    public static MutationSettingsData MutationsTable {
        get {
            if (mutationsTable == null) {
                mutationsTable = Resources.Load<MutationSettingsData>(MutationTablePath);
            }
            return mutationsTable;
        }
    }

    public MutationManager(Unit unit, Battle battle) {
        this.unit = unit;
        this.battle = battle;
        hits = new Dictionary<MutantEvent, int>();
        level = battle.Enemy.HighestMeatLevel;
    }

    public void RecordEvent(MutantEvent @event) {
        var existing = 0;
        hits.TryGetValue(@event, out existing);
        hits[@event] = existing + 1;
    }

    public List<Mutation> GenerateOptions() {
        var occurred = new List<MutantEvent>();
        foreach (var @event in hits) {
			if (@event.Value > 0) {
                occurred.Add(@event.Key);
            }
        }

        Mutation @fixed = null;
		if (occurred.Count > 0) {
            var randomOccurred = occurred[Random.Range(0, occurred.Count)];
			switch (randomOccurred) {
			    case MutantEvent.DAMAGED:
				    @fixed = GenerateHealthOption();
				    break;
			    case MutantEvent.DAMAGED_PHYSICALLY:
				    @fixed = new StatMutation(unit, StatTag.DEF);
				    break;
			    case MutantEvent.USED_ABILITY:
				    @fixed = new AbilMutation(unit, level);
				    break;
			    case MutantEvent.USED_AGI:
				    @fixed = new StatMutation(unit, StatTag.AGI);
				    break;
			    case MutantEvent.USED_MANA:
				    @fixed = new StatMutation(unit, StatTag.MANA);
				    break;
			    case MutantEvent.USED_STR:
				    @fixed = new StatMutation(unit, StatTag.STR);
				    break;
			}
		} else {
			@fixed = GenerateRandomOption();
		}
		
		var results = new List<Mutation>();
		while (IsMaxed(@fixed)) {
			@fixed = GenerateRandomOption();
		}
		results.Add(@fixed);
		
		for (var tries = 0; tries < 100; tries += 1) {
            var random = GenerateRandomOption();
			if (@fixed.Stat != random.Stat && !IsMaxed(random)) {
				results.Add(random);
				break;
			}
		}
		if (results.Count < 2) {
			results.Add(new AbilMutation(unit, level));
		}
		
		return results;
	}
	
	protected bool IsMaxed(Mutation mutation) {
        var stat = mutation.Stat?.Tag;
        if (stat.HasValue) {
            if (stat == StatTag.MHP) {
                return unit.BaseStats[stat.Value] >= 999;
            } else {
                return unit.BaseStats[stat.Value] >= 99;
            }
        } else {
            return false;
        }
    }

    private Mutation GenerateHealthOption() {
        var gain = Random.Range(MhpGainMin, MhpGainMax);
        return new StatMutation(unit, StatTag.MHP, gain);
    }

    private Mutation GenerateRandomOption() {
        var spread = new List<MutantEvent>();
        for (var i = 0; i < MutationsTable.weightHp; i += 1) {
            spread.Add(MutantEvent.DAMAGED);
        }
        for (var i = 0; i < MutationsTable.weightDef; i += 1) {
            spread.Add(MutantEvent.DAMAGED_PHYSICALLY);
        }
        for (var i = 0; i < MutationsTable.weightAbil; i += 1) {
            spread.Add(MutantEvent.USED_ABILITY);
        }
        for (var i = 0; i < MutationsTable.weightAgi; i += 1) {
            spread.Add(MutantEvent.USED_AGI);
        }
        for (var i = 0; i < MutationsTable.weightStr; i += 1) {
            spread.Add(MutantEvent.USED_STR);
        }
        for (var i = 0; i < MutationsTable.weightMana; i += 1) {
            spread.Add(MutantEvent.USED_MANA);
        }
        switch (spread[Random.Range(0, spread.Count)]) {
            case MutantEvent.DAMAGED:               return GenerateHealthOption();
            case MutantEvent.DAMAGED_PHYSICALLY:    return new StatMutation(unit, StatTag.DEF);
            case MutantEvent.USED_ABILITY:          return new AbilMutation(unit, level);
            case MutantEvent.USED_AGI:              return new StatMutation(unit, StatTag.AGI);
            case MutantEvent.USED_MANA:             return new StatMutation(unit, StatTag.MANA);
            case MutantEvent.USED_STR:              return new StatMutation(unit, StatTag.STR);
            default: return null;
        }
    }
}
